lib/ansible/modules/database/postgresql/postgresql_query.py

```diff
@@ -0,0 +1,330 @@
+    query: SELECT * FROM some_table WHERE a_column=%(a_value)s AND b_column=%(b_value)s
+    query: /path/to/my/script.sql
+    named_args:
+      a_value: "positional string value 1"
+      b_value: "positional string value 2"
+    fact: my_key
+'''
+
+RETURN = '''
+query_results:
+    description: List of dictionaries in column:value form representing returned rows.
+    returned: changed
+    type: list
+    sample: [{"Column": "Value1"},{"Column": "Value2"}]
+row_count:
+    description: number of affected rows by query, if applicable
+    returned: changed
+    type: int
+    sample: 5
+'''
+
+import traceback
+from ansible.module_utils.six import iteritems
+from ansible.module_utils.basic import AnsibleModule, BOOLEANS
+import ansible.module_utils.postgres as pgutils
+
+
+def main():
+    argument_spec = pgutils.postgres_common_argument_spec()
+
+    argument_spec.update(dict(
+        db=dict(default=None),
+        query=dict(type="str"),
+        autocommit=dict(type='bool', default=False, choices=BOOLEANS),
+        positional_args=dict(type="list"),
+        named_args=dict(type="dict"),
+        fact=dict(default=None),
+    ))
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        mutually_exclusive=[
+            ["positional_args", "named_args"]
+        ],
+        supports_check_mode=False,
+    )
+
+    changed = False
+    psycopg2 = pgutils.psycopg2
+    ansible_facts = {}
+    query_results = []
+    rowcount = 0
+
+    # To use defaults values, keyword arguments must be absent, so
+    # check which values are empty and don't include in the **kw
+    # dictionary
+    params_map = {
+        "login_host": "host",
+        "login_user": "user",
+        "login_password": "password",
+        "port": "port",
+        "db": "database",
+        "ssl_mode": "sslmode",
+        "ssl_rootcert": "sslrootcert"
+    }
+
+    kw = dict((params_map[k], v) for (k, v) in iteritems(module.params)
+              if k in params_map and v != '' and v is not None)
+
+    # If a login_unix_socket is specified, incorporate it here.
+    is_localhost = "host" not in kw or kw["host"] == "" or kw["host"] == "localhost"
+
+    if is_localhost and module.params["login_unix_socket"] != "":
+        kw["host"] = module.params["login_unix_socket"]
+
+    try:
+        pgutils.ensure_libs(sslrootcert=module.params.get('ssl_rootcert'))
+        db_connection = psycopg2.connect(**kw)
+
+        # Enable autocommit on demand only so we can create databases
+        if module.params['autocommit']:
+            if psycopg2.__version__ >= '2.4.2':
+                db_connection.autocommit = True
+            else:
+                db_connection.set_isolation_level(psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT)
+
+        # Using RealDictCursor allows access to row results by real column name
+        cursor = db_connection.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+
+    except TypeError as e:
```

Comment:
- `TypeError` won't be raised: `ensure_libs` raise `LibraryError`.
- There is no need to differentiate `LibraryError` and `Exception`. 
- `db_connection.close()` must called before `module.fail_json()` (`fail_json` calls `sys.exit` nothing is executed after)

 It seems the `try` block could be updated:
```
try:
    ensure_libs(...)
    [...]
except Exception as e:
    db_connection.close()
    module.fail_json(msg=to_native(e), exception=traceback.format_exc())
```

Reply:
I think `TypeError` can be catched as it is placed before `Exception` that is less specific. 

I agree there is no need to differentiate `LibraryError` and `Exception`. I think we need to leave this be for the time beeing inside `postgres.py` until we harmonize all postgres modules further more. i fixed the `db_connection.close()` already.

Reply:
Two things:
- `ensure_libs(...)` raise `LibraryError` when `sslrootcerts` isn't supported: should not `TypeError` be replaced with `LibraryError` ? If not which statement might raise a `TypeError` exception ?
- `ensure_libs()` raises an error telling that [_psycopg2 must be at least 2.4.3_](https://github.com/archf/ansible/blob/bf17b7df2db50ac4c91816ec3737c0c2c35dc994/lib/ansible/module_utils/postgres.py#L45) and this module states that `Postgresql server must be at least version 8.4 to support sslrootcert`. In order to use [`sslrootcert`](https://www.postgresql.org/docs/8.4/static/libpq-connect.html#LIBPQ-CONNECT-SSLROOTCERT), `libpq` version should be at least 8.4 but PostgreSQL could be older. `sslrootcert` feature doesn't require a minimum PostgreSQL server version. It seems exception message raised by `ensure_libs` should be `sslrootcert: psycopg2 >= 2.4.3 and libpq >= 8.4 required` (it should be fixed in another pull request). That's why I would prefer only one `except` block there.

Reply:
ok, i removed it. Postgresql 8.X is not something i really care about. I was just not brave enough to try install Postgres that old and fiddle with it so i left it there and did not question to much. Might be on RHEL 5.X... Anyhow we are still catching an exception...